src/algebra/lie/direct_sum.lean

```diff
@@ -87,27 +89,121 @@ instance : lie_ring (⨁ i, L i) :=
+`(⨁ i, L i) →ₗ[R] L'`. If in addition `⁅f i x, f j y⁆ = 0` for any `x ∈ L i` and `y ∈ L j` (`i ≠ j`)
+then this map is a morphism of Lie algebras. -/
+@[simps] def to_lie_algebra [decidable_eq ι] (L' : Type w₁) [lie_ring L'] [lie_algebra R L']
+  (f : Π i, L i →ₗ⁅R⁆ L') (hf : ∀ (i j : ι), i ≠ j → ∀ (x : L i) (y : L j), ⁅f i x, f j y⁆ = 0) :
+  (⨁ i, L i) →ₗ⁅R⁆ L' :=
+{ to_fun   := to_module R ι L' (λ i, (f i : L i →ₗ[R] L')),
+  map_lie' := λ x y,
+    begin
+      let f' := λ i, (f i : L i →ₗ[R] L'),
+      /- The goal is linear in `y`. We can use this to reduce to the case that `y` has only one
+        non-zero component. -/
+      suffices : ∀ (i : ι) (y : L i), to_module R ι L' f' ⁅x, lof R ι L i y⁆ =
+        ⁅to_module R ι L' f' x, to_module R ι L' f' (lof R ι L i y)⁆,
+      { simp only [← lie_algebra.ad_apply R, ← linear_map.comp_apply],
+        congr, clear y, ext i y, exact this i y, },
+      /- Similarly, we can reduce to the case that `x` has only one non-zero component. -/
+      suffices : ∀ i j (y : L i) (x : L j), to_module R ι L' f' ⁅lof R ι L j x, lof R ι L i y⁆ =
+        ⁅to_module R ι L' f' (lof R ι L j x), to_module R ι L' f' (lof R ι L i y)⁆,
+      { intros i y,
+        rw [← lie_skew x, ← lie_skew (to_module R ι L' f' x)],
+        simp only [linear_map.map_neg, neg_inj, ← lie_algebra.ad_apply R, ← linear_map.comp_apply],
+        congr, clear x, ext j x, exact this j i x y, },
+      /- Tidy up and use `lie_lof`. -/
+      intros i j y x,
+      simp only [to_module_lof, lie_algebra_of_apply, lie_hom.coe_to_linear_map, lie_lof],
+      /- And finish with trivial case analysis. -/
+      rcases eq_or_ne i j with h | h,
+      { have h' : f j (h.rec_on y) = f i y, { exact eq.drec (eq.refl _) h, },
+        simp only [h, to_module_lof, lie_hom.coe_to_linear_map, dif_pos, eq_self_iff_true,
+          lie_hom.map_lie, h'], },
+      { simp only [h, hf j i h.symm x y, linear_map.map_zero, dif_neg, not_false_iff], },
+    end,
+.. to_module R ι L' (λ i, (f i : L i →ₗ[R] L')) }
+
 end algebras
 
+section ideals
+
+variables {L : Type w} [lie_ring L] [lie_algebra R L] (I : ι → lie_ideal R L)
+
+/-- Given a Lie algebra `L` and a family of ideals `I i ⊆ L`, informally this definition is the
+statement that `L = ⨁ i, I i`.
+
+More formally, the inclusions give a natural map from the (external) direct sum to the enclosing Lie
+algebra: `(⨁ i, I i) → L`, and this definition is the proposition that this map is bijective. -/
+def lie_algebra_is_internal [decidable_eq ι] : Prop :=
+function.bijective $ to_module R ι L $ λ i, ((I i).incl : I i →ₗ[R] L)
+
+/-- The fact that this instance is necessary seems to be a bug in typeclass inference. See
+[this Zulip thread](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/
+Typeclass.20resolution.20under.20binders/near/245151099). -/
+instance lie_ring_of_ideals : lie_ring (⨁ i, I i) := direct_sum.lie_ring (λ i, ↥(I i))
+
+/-- See `lie_ring_of_ideals` comment. -/
```

Comment:
This needs to be a fully-qualified name for doc-gen to link it.

Reply:
Thanks!